### DISCUSSION TITLE
.find() with populate uses the wrong fieldNames

### DISCUSSION THREAD

**Discussion by quazeenii** [mikro-orm/mikro-orm#3738] [Q&A]
Hi,

I have tried searching though I wasn't sure how to word it exactly, I apologize if this was already discussed. I believe my issues is similar to [this](https://github.com/mikro-orm/mikro-orm/issues/3669) one.

I have two entities, e.g. Questions and Answers.

```
@Entity({ tableName: 'questions' })
export class Question {
  [PrimaryKeyType]?: [string, Date];

  @PrimaryKey({ columnType: 'uuid', defaultRaw: `gen_random_uuid()` })
  id: UUID = UUID();

  @PrimaryKey({ length: 6, defaultRaw: `now()` })
  createdAt: Date = new Date();

  @Property({ length: 6, defaultRaw: `now()`, onUpdate: () => new Date() })
  updatedAt: Date = new Date();

  @OneToMany(() => Answer, (answer) => answer.question)
  answers: Collection<Answer> = new Collection<Answer>(this);

  @Property({ length: 255 })
  name: string;
}
```

```
@Entity({ tableName: 'answers' })
export class Answer {
  [PrimaryKeyType]?: [string, Date];

  @PrimaryKey({ columnType: 'uuid', defaultRaw: `gen_random_uuid()` })
  id: UUID = UUID();

  @PrimaryKey({ length: 6, defaultRaw: `now()` })
  createdAt: Date = new Date();

  @Property({ length: 6, defaultRaw: `now()`, onUpdate: () => new Date() })
  updatedAt: Date = new Date();

  @ManyToOne({ entity: () => Question })
  question: Question;
}
```

in the db, `Answer.question` actually maps to `answers.question_id` and `answers.question_created_at` since questions has a multi-pkey that answers need to store

finding many fails when I do a find with the populate,
```
await this.em.find(Answer, { question }, { populate: ['question'] })
```
because the where clause looks for `questions.question_id` and `questions.question_created_at` instead of either `questions.id` and `questions.created_at` or `answers.question_id `and `answers.question_created_at`
```
where ("questions"."question_id", "questions"."question_created_at") = ('0956160c-87cf-44d3-b27e-9b24034d1c24', '2022-11-13T05:00:39.994Z') order by "answers"."created_at" desc
```

but it works without the populate (question just isn't populated)
```
await this.em.find(Answer, { question })
```
and it works if I use queryBuilder instead
```
const answers = await this.em.createQueryBuilder(Answer).where({question}).getResult()
await em.populate(answers, ['question'])
```

I was wondering if I was doing something wrong or if I'm supposed to just use queryBuilder?
is one better than the other? queryBuilder vs find?

**Comment by B4nan** [comment]
Can you compile this into a complete script I can run myself? As in https://github.com/mikro-orm/mikro-orm/tree/master/tests/issues

> but it works without the populate (question just isn't populated)

The query that fails is the populate query - no populate hint -> no populate queries. If you do `await em.populate(answers, ['question'])` that would be equivalent and it will probably fail the same.

Generally speaking, I would discourage this - it only adds complexity to the entities, your UUID is unique on its own, so you don't need a composite PK here. On the other hand, its a matter of preference, and it should work, especially if it works with the QB + `em.populate`.

edit: now I see you do `await this.em.find(Answer, { question }, { populate: ['Question'] })`, that populate hint is wrong. you need to put there a property name, not entity name. this wont even type check. but I guess thats a typo in your code here, rather than in your actual app?

**Reply by quazeenii** [reply]
yes sorry a typo. fixed in the question.

which part exactly are you discouraging?

the part that is failing is the where query when adding `populate` in with the `.find()`

when I break it into queryBuilder and then populate, it works because the where query is correct, and then it populates all the entities after (I assume).

I will create the script at some point today and attach.

**Reply by B4nan** [reply]
having composite keys formed from already unique columns

**Reply by quazeenii** [reply]
oh yes, but since we're using partitions, apparently the partition key needs to be part of the pkey

is there a way for me to run the scripts to make sure I have it working properly for you to test?

**Reply by quazeenii** [reply]
[github_test.txt](https://github.com/mikro-orm/mikro-orm/files/10004800/github_test.txt)

here's the file. I figured out the issue, it's because I have `LoadStrategy.JOINED` instead of `LoadStrategy.SELECT_IN`

I also found another thing I wasn't expecting (the commented out questions in the tests), `question` is always returned in `answers`. is that because of the LoadStrategy? is that expected? do I even need the populate field if it's always returned?

Thanks!

**Reply by B4nan** [reply]
All bidirectional relations propagate their changes, so if you populate one side, the other will get populated too. When you serialize your entities, the populate hints should be respected, so this "auto-populated" relations shouldn't be expanded as if they were populated explicitly.

https://mikro-orm.io/docs/propagation
https://mikro-orm.io/docs/identity-map

**Reply by B4nan** [reply]
Or well, why do you think it should be even undefined? The property represents a real database column, the foreign key - the value is there even without population. It will be represented as an entity instance.

https://mikro-orm.io/docs/next/entity-manager#entity-references

**Reply by quazeenii** [reply]
oh no, I just wasn't expecting it to be there since I didn't have the populate, I guess I assumed you would just ignore the relations. your links make sense I think, I'll have to look at it more in a bit.

but then I think there's still an issue when using LoadStrategy.JOINED which is shown in the test. unless you don't think so, then can you explain why and what I should do instead?

**Reply by B4nan** [reply]
Yes, it is indeed valid, looking into it now.

**Reply by B4nan** [reply]
Here you go, it was easier than expected. Dev version will be available in few minutes.

https://github.com/mikro-orm/mikro-orm/commit/dec4c9c46b1ecf3105f78b77a698c30ef8670c14

**Reply by quazeenii** [reply]
thanks!
